### DISCUSSION TITLE
How to do Build-time imports without PostCSS.

### DISCUSSION THREAD

**Discussion by tordans** [tailwindlabs/tailwindcss#6950] [Help] [answered]
Alternative title: **Rails 7 & TailwindCSS JIT: Split css in separate files, compile one output**

I am using the new Rails 7 + TailwindCSS JIT setup. And I have a `application.tailwind.css` files that gets compiled via the TailwindCLI to an `application.css` file. 

I want this file to be split up into multiple files, but still be able to use all the TailwindCSS-niceness inside the sub-files.
Example:
- `stylesheets/application.tailwind.css` would hold the `@tailwind` directives and some general `@layer` with classes using `@apply`
- `stylesheets/buttons.css` would only hold `@layer components {.btn { @apply … }}` and such

In https://github.com/tailwindlabs/tailwindcss/discussions/6612 and https://tailwindcss.com/docs/using-with-preprocessors#build-time-imports Adam explains a solution using PostCSS with `postcss-import`. However, one of the great things about Rails 7 + TailwindCSS is to not have PostCSS anymore.

So what is the non-PostCSS-Solution for this?

**Comment by Xeevis** [comment]
You mean you are using TailwindCSS CLI? What is wrong with PostCSS? It combines multiple files into a single file and generates sourcemaps that map classes to their individual source files. What exactly is your problem, you don't want Node installed?

**Reply by tordans** [reply]
Yes, I am using the CLI which is the new default setup for Ruby on Rails 7. One core goal of Rails 7 is, not to have PostCSS as part of the dependency mix. That is why I am looking for a way to achieve what is now documented to be solved by a PostCSS-Solution in a different way.

For reference:
- Adam promoting using the CLI over tighly integrated solutions https://twitter.com/adamwathan/status/1467196301238542340?s=12
- DHH promoting Rails7+TW without Node https://twitter.com/dhh/status/1472002099684659202
- The new intro video on https://rubyonrails.org/ gives a good idea about the new default setup for Rails7

**Comment by tordans** [comment]
Reading the docs at https://remix.run/docs/en/v1/guides/styling#postcss (after writing my answer https://github.com/tailwindlabs/tailwindcss/discussions/6950#discussioncomment-1931474) I wonder if the new setup should look like this:

1. I use PostCSS to gather all `/css-step1/*.css` into one `/css-step2/application.tailwind.css`
1. I use the tailwindCLI to process the `/css-step2/tailwind.input.css` into a `/app/assets/builds/application.css`
1. Rails then picks up this file from there

**Reply by ManuelFrigerio** [reply]
Did you ever sort this out?

**Reply by tordans** [reply]
No, I think it is not solved, yet. I guess the stand alone tw cli would need additional feature but there is no dedicated place to report this need afaik.

**Comment by Xeevis** [comment]
@ManuelFrigerio @tordans I have solved this by moving away from Node.js to it's younger brother [Deno](https://deno.land). Deno is a single executable JS/TS runtime that has no other dependencies, modules are cached directly from URL and is compatible with Node packages trough polyfills which is done automatically by esm.sh CDN. So basically all I need are **tailwindcss.exe**, **deno.exe** and **build.ts**. In development I start 2 watchers on separate threads, one Deno watcher that is using native OS API and watches all pcss/ts/static files for changes. If changed file is `*.pcss` I call PostCSS to build **app.pcss** with imports to a single **tempfile.css**. And another watcher from TailwindCSS CLI with `--watch` argument that watches this `tempfile.css` and builds `app.css`. For TS files I bundle them with esbuild and static files i copy (mirror) with `robocopy`, Production build will use slightly different flow, I use esbuild to minify css instead of cssnano that is bundled with tailwindCSS CLI,

Here is some dirty code that most likely won't run properly I just gutted out of a bit more complex flow in the middle of rework just to give you an idea how this works with postcss, postcss-import and nesting. Basically you then just run `deno --allow-all build.ts --watch`, you might even [compile](https://deno.land/manual/tools/compiler) your build flow into standalone executable.

```ts
// Standard packages
import { parse } from "https://deno.land/std@0.123.0/flags/mod.ts";
import { expandGlob } from "https://deno.land/std@0.123.0/fs/mod.ts";
import { extname, join } from "https://deno.land/std@0.123.0/path/mod.ts";
import { debounce } from "https://deno.land/std@0.123.0/async/mod.ts";

// NPM packages
import postcss, { AcceptedPlugin } from "https://esm.sh/postcss@8.4.5";
import postcss_import from "https://esm.sh/postcss-import@14.0.2";
import postcss_nesting from "https://esm.sh/tailwindcss@3.0.18/nesting/index.js";

const args = parse(Deno.args);
const production = args["production"];
const watch = args["watch"];

// VS Code might save files twice, debouncer makes sure we process only one such event
const eventDebouncer = debounce(async (event: Deno.FsEvent) => {
  const path = event.paths[0] as string;
  console.log("Detected [%s] event for %s", event.kind, path);
  switch (extname(path)) {
    case ".pcss":
      await transformPostCSS();
      break;
    case ".ts":
      //await transformTS(path);
      break;
    default:
      //await syncStaticFiles();
      break;
  }
}, 100);

// First we build everything
await Promise.all([
  (async () => {
    await transformPostCSS();
  })(),
  (async () => {
    //await transformTS();
  })(),
  (async () => {
    //await syncStaticFiles();
  })(),
]);

if (watch) {
  console.log("Starting watchers ...");
  (async () => {
    await transformTailwindCSS();
  })();
  (async () => {
    for await (const event of Deno.watchFs([
      "assets",
      "Pages",
      "Shared",
      "App.razor",
    ])) {
      eventDebouncer(event);
    }
  })();
}

const postCssTempFile = await Deno.makeTempFile();

async function transformPostCSS() {
  const result = await postcss([
    postcss_import({
      // add support for importing globs in pcss
      async resolve(id: string, base: string) {
        const files: Array<string> = [];
        for await (const file of expandGlob(join(base, id))) {
          files.push(file.path);
        }
        return files;
      },
    }) as AcceptedPlugin,
    postcss_nesting as AcceptedPlugin,
  ]).process(await Deno.readTextFile("assets/css/app.pcss"), {
    from: "assets/css/app.pcss",
    map: !production
  });

  await Deno.writeTextFile(postCssTempFile, result.css);
}

async function transformTailwindCSS() {
  const cmdArgs = [
    "cmd",
    "/C",
    "tailwindcss",
    "-i",
    postCssTempFile,
    "-c",
    "tailwind.config.cjs",
  ];

  if (!production) {
    cmdArgs.push("-o", "wwwroot/css/app.css");
  }

  if (watch) {
    cmdArgs.push("--watch");
  }

  const twprocess = Deno.run({
    cmd: cmdArgs,
    stdin: "piped",
    stdout: "piped",
  });

  if (!watch) {
    await twprocess.stdin.write(
      new TextEncoder().encode(await Deno.readTextFile(postCssTempFile))
    );
    twprocess.stdin.close();
  }

  const [_status, stdout] = await Promise.all([
    twprocess.status(),
    twprocess.output(),
  ]);
  twprocess.close();

  if (minify) {
    await minifyCss(new TextDecoder().decode(stdout));
  }
}

async function minifyCss(css: string) {
  console.log("Minifying with esbuild ...");
  await esbuild.build({
    stdin: { contents: css, loader: "css" },
    minify: true,
    outfile: "wwwroot/css/app.css",
  });
  esbuild.stop();
}
```

**Comment by tordans** [comment, marked as answer]
This is solved in Tailwind CSS 3.1. At least according to https://youtu.be/nOQyWbPO2Ds 2:36. The stand alone CLI has now the ability to include css files that where split from the main file or external.

**Comment by dfallman** [comment]
Played around with this for a project just now and I think the instructions/docs aren't 100% clear (or could be just me too of course). 

The way it seems to work for me is to do this in `main.css`:

```
@import "tailwindcss/base";
@import "./css/tweaks.css";
@import "tailwindcss/components";
@import "tailwindcss/utilities";

@layer components {
	.focus-normal {
		@apply focus:outline-none focus:ring-0 focus-visible:ring-2 focus-visible:ring-blue-600;
	}
}
```
Where `tweaks.css` is a custom stylesheet with additional css required by some components.

**Comment by joeyame** [comment]
Anyone who runs into this in the future, I had:
```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@import "./markdown.css";
```

and the contents of `markdown.css` never made it into the output file. Switching the standard tailwind imports over to what the previous comment does fixed the problem!

```css
@import "tailwindcss/base";
@import "tailwindcss/components";
@import "tailwindcss/utilities";

@import "./markdown.css";
```
